### PR TITLE
investigated improvements for requester publish with same correlationId

### PR DESCRIPTION
**Pull request by rkatsyuryna**

### CODE REVIEW

File: core/src/main/java/io/github/tcdl/msb/api/ObjectFactory.java [hunk 1]
```diff
@@ -18,16 +17,6 @@
...
 
     /**
-     * Creates a new instance of {@link Requester} with originalMessage.
-     *
-     * @param namespace       topic name to send a request to
-     * @param requestOptions  options to configure a requester
-     * @param originalMessage original message (to take correlation id from)
-     * @return instance of a requester
-     */
-    Requester createRequester(String namespace, RequestOptions requestOptions, Message originalMessage);
```
Comment (anstr) [inline core/src/main/java/io/github/tcdl/msb/api/ObjectFactory.java:28]: I'm not sure that  removing method createRequester which uses originalMessage as parameter is good idea.

File: examples/src/main/java/io/github/tcdl/msb/examples/FacetsAggregator.java [hunk 2]
```diff
@@ -86,7 +86,7 @@ public static void main(String[] args) throws ScriptException, FileNotFoundExcep
                     responder.send(responsePayload);
                 });
 
-                requester.publish(request);
+                requester.publish(request, responder.getOriginalMessage().getCorrelationId());
```
Comment (anstr) [inline examples/src/main/java/io/github/tcdl/msb/examples/FacetsAggregator.java]: I think that author of microservices should know about correlationId,as for me previous approach when we keep correlationId inside ResponderImpl is better.

File: core/src/main/java/io/github/tcdl/msb/api/ObjectFactory.java (hunk 1 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/msb/api/ObjectFactory.java:28]: as we discussed we agree on it

File: core/src/main/java/io/github/tcdl/msb/message/MessageFactory.java [hunk 3]
```diff
@@ -58,34 +39,46 @@ public Message createResponseMessage(Message.Builder messageBuilder, Acknowledge
...
     }
 
-    private Message.Builder createBaseMessage(@Nullable Message originalMessage) {
-        Message.Builder baseMessage = new Message.Builder()
-                .withId(Utils.generateId())
-                .withCorrelationId(
-                        originalMessage != null && originalMessage.getCorrelationId() != null ? originalMessage
-                                .getCorrelationId() : Utils.generateId());
+    private Builder createMetaBuilder(MessageTemplate config) {
+        Integer ttl = config == null ? null : config.getTtl();
```
Comment (anstr) [inline core/src/main/java/io/github/tcdl/msb/message/MessageFactory.java]: Please put ternary operator inside brackets.

File: core/src/main/java/io/github/tcdl/msb/api/Requester.java [hunk 4]
```diff
@@ -31,6 +31,17 @@
...
+    /**
+     * Wraps a payload with protocol information, preserves original message correlationId and sends to bus.
+     * In case Requester created with expectation for responses then process them.
+     *
+     * @param requestPayload payload which will be sent to bus
+     * @param correlationId payload which will be sent to bus
+     * @throws ChannelException if an error is encountered during publishing to bus
+     * @throws JsonConversionException if unable to parse message to JSON before sending to bus
+     */
+    void publish(Payload requestPayload, String correlationId);
```
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/msb/api/Requester.java]: We should have message here. correlationId in only one possible field that get propagated.

File: examples/src/main/java/io/github/tcdl/msb/examples/FacetsAggregator.java (hunk 2 above)
Comment (rkatsyuryna) [inline examples/src/main/java/io/github/tcdl/msb/examples/FacetsAggregator.java]: what means inside? we pass this to Requester constructor. now we pass this to publish method.

File: core/src/main/java/io/github/tcdl/msb/message/MessageFactory.java (hunk 3 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/msb/message/MessageFactory.java]: sorry what part of code is that about? createMetaBuilder was there for ages like it is

File: core/src/main/java/io/github/tcdl/msb/api/Requester.java (hunk 4 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/msb/api/Requester.java]: agree. pass Message is fine as well, but for future API update i guess

File: core/src/main/java/io/github/tcdl/msb/api/exception/DuplicateCollectorException.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package io.github.tcdl.msb.api.exception;
+
+/**
+ * Exception is thrown in case more then one Collector is created per correlationId.
+ */
+public class DuplicateCollectorException extends MsbException {
```
Comment (anstr) [inline core/src/main/java/io/github/tcdl/msb/api/exception/DuplicateCollectorException.java:6]: Maybe CollectorSameCorrelationIdException will be more suitable?

File: core/src/main/java/io/github/tcdl/msb/message/MessageFactory.java (hunk 3 above)
Comment (anstr) [inline core/src/main/java/io/github/tcdl/msb/message/MessageFactory.java]: Yes it was,but i think that brackets will make that code more readable.

File: core/src/main/java/io/github/tcdl/msb/api/exception/DuplicateCollectorException.java (hunk 5 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/msb/api/exception/DuplicateCollectorException.java:6]: two manu subjects in name. Not good Practice

**Pull request merged**